yarn-project/foundation/src/json-rpc/server/safe_json_rpc_server.ts

```diff
@@ -0,0 +1,243 @@
+import cors from '@koa/cors';
+import http from 'http';
+import Koa from 'koa';
+import bodyParser from 'koa-bodyparser';
+import compress from 'koa-compress';
+import Router from 'koa-router';
+import { format } from 'util';
+import { ZodError } from 'zod';
+
+import { createDebugLogger } from '../../log/index.js';
+import { promiseWithResolvers } from '../../promise/utils.js';
+import { type ApiSchema, type ApiSchemaFor, schemaHasMethod } from '../../schemas/index.js';
+import { jsonStringify2 } from '../convert.js';
+import { assert } from '../js_utils.js';
+
+export class SafeJsonRpcServer {
+  /**
+   * The HTTP server accepting remote requests.
+   * This member field is initialized when the server is started.
+   */
+  private httpServer?: http.Server;
+
+  constructor(
+    /** The proxy object to delegate requests to. */
+    private readonly proxy: Proxy,
+    /** Logger */
+    private log = createDebugLogger('json-rpc:server'),
+  ) {}
+
+  /**
+   * Get an express app object.
+   * @param prefix - Our server prefix.
+   * @returns The app object.
+   */
+  public getApp(prefix = '') {
+    const router = this.getRouter(prefix);
+
+    const exceptionHandler = async (ctx: Koa.Context, next: () => Promise<void>) => {
+      try {
+        await next();
+      } catch (err: any) {
+        this.log.error(err);
+        if (err instanceof SyntaxError) {
```

Comment:
One wishlist item for me would be a request ID that is used to log this error so that it can be easily correlated from the tx bot e.g. side

Reply:
Not just for you: https://hackmd.io/@aztec-network/rkebFFEhA?type=view#Tracing-requests